### ISSUE TITLE
Write string EXIF-tags with libvips

### ISSUE THREAD

**Issue by goarsna** [kleisauke/net-vips#10]
Hi,

I've got a problem with writing meta data to a JPEG file. I suppose this must be because of a datatype mismatch.

Here's a little example program, where I experience this error: 
```
var vipsImage = NetVips.Image.NewFromFile("test.jpg"); // Load iamge

System.Console.WriteLine("Old orientation: " + vipsImage.Get("orientation"));
vipsImage.Set("orientation", 2); // Change orientation
System.Console.WriteLine("New orientation: " + vipsImage.Get("orientation"));

System.Console.WriteLine("Old key words: " + vipsImage.Get("exif-ifd0-XPKeywords"));
vipsImage.Set("exif-ifd0-XPKeywords", "test keyword"); // Change key words
System.Console.WriteLine("New key words: " + vipsImage.Get("exif-ifd0-XPKeywords"));

vipsImage.Jpegsave("result.jpg"); // Write to file

var reloadedImage = NetVips.Image.NewFromFile("result.jpg"); // Reload file

System.Console.WriteLine("Orientation after reload: " + reloadedImage.Get("orientation"));
System.Console.WriteLine("Key words after reload: " + vipsImage.Get("exif-ifd0-XPKeywords"));

```
(Note: For the above code snippet to work, the Tags field in **Properties => Details** of the loaded JPEG file must at least have one tag set. Otherwise, of course, **vipsImage.Get("exif-ifd0-XPKeywords")** will fail)

and the corresponding output:
```
Old orientation: 1
New orientation: 2
Old key words: Test-Tag (Test-Tag, Byte, 18 components, 18 bytes)
New key words: test keyword
Orientation after reload: 2
Key words after reload: Test-Tag (Test-Tag, Byte, 18 components, 18 bytes)
```

As you can see, I'm able to set a new **orientation** in meta data and after saving the file to disk (doesn't matter if I use Jpegsave or WriteToFile) and reloading the saved file again, the **orientation** still has my changed value but the **exif-ifd0-XPKeywords** field doesn't.
Now my guess is, this behavior is caused by a type mismatch of the **exif-ifd0-XPKeywords** field and the string i try to save into it. But the only hint I have is the different output of "Old keywords" and "New key words". Looking at the byte information in the "Old key words"-output i suppose i have to save the "test keyword"-string in a different way. But I don't know how. And also i don't know why, because before saving the image to disc, the changes seem to have an effect.

By debugging I saw that the type of this field is **GValue.RefStrType** and the part of the method **GValue.Set()** for **GValue.RefStrType** equals the part for **GValue.GStrType** while the **GValue.Get()** method differs at this point.

Now my question: Is this an error in NetVips or is the error in front of my keyboard? :)

I hope my problem is understandable and I'm thankful for any hint on how to properly save key words in the exif data.
If you need any further information, don't hesitate to ask.

Thank You

**Comment by kleisauke** [comment]
Hi,

I could reproduce this. It looks like it's impossible to set these Microsoft EXIF fields:
```
exif-ifd0-XPTitle
exif-ifd0-XPComment
exif-ifd0-XPAuthor
exif-ifd0-XPKeywords
exif-ifd0-XPSubject
```

@jcupitt, Is it possible that this is a character encoding issue? I noticed that these [Microsoft EXIF fields](https://github.com/libexif/libexif/blob/80ed9dffc0666246a4852936443846d24609aef4/libexif/exif-tag.c#L643-L667) needs to be encoded as `UTF-16LE`.

I've made a reproducible test case in pyvips:
```python
#!/usr/bin/python

import pyvips

# https://t0.nl/exif-test.jpg
image = pyvips.Image.new_from_file('exif-test.jpg')

print('Old keywords: ' + image.get('exif-ifd0-XPKeywords'))
new_image = image.copy()
new_image.set('exif-ifd0-XPKeywords', 'test;keyword')
print('New keywords: ' + new_image.get('exif-ifd0-XPKeywords'))

new_image.write_to_file('result.jpg')

reloaded_image = pyvips.Image.new_from_file('result.jpg')

print('Keywords after reload: ' + reloaded_image.get('exif-ifd0-XPKeywords'))
```

Output:
```
Old keywords: one;two;three (one;two;three, Byte, 28 components, 28 bytes)
New keywords: test;keyword
Keywords after reload: one;two;three (one;two;three, Byte, 28 components, 28 bytes)
```

**Comment by jcupitt** [comment]
Oh, interesting problem, I'll have a look. It can read the tags from your sample image correctly, on linux at least, though I'm not quite sure how:

```
john@kiwi:~/pics$ vipsheader -a exif-test.jpg  | grep -i xpcomm
exif-ifd0-XPComment: Comment (Comment, Byte, 16 components, 16 bytes)
```

I guess libexif is doing the utf16 -> utf8 automatically on read.

**Comment by jcupitt** [comment]
Oh heh libvips will only save numeric values to exif -- I guess I just never needed to set strings. See:

https://github.com/jcupitt/libvips/blob/master/libvips/foreign/exif.c#L880

```python
#!/usr/bin/env python

import sys
import pyvips

# https://t0.nl/exif-test.jpg
image = pyvips.Image.new_from_file(sys.argv[1])

tag = 'exif-ifd0-ImageDescription'

print('Old {0} = {1}'.format(tag, image.get(tag)))
new_image = image.copy()
new_image.set(tag, 'test;keyword')
print('New {0} = {1}'.format(tag, new_image.get(tag)))

new_image.write_to_file('x.jpg')

reloaded_image = pyvips.Image.new_from_file('x.jpg')

print('After reload {0} = {1}'.format(tag, reloaded_image.get(tag)))
```

I see:

```
john@kiwi:~/try$ ./exif2.py ~/pics/exif-test.jpg 
Old exif-ifd0-ImageDescription = Title (Title, ASCII, 6 components, 6 bytes)
New exif-ifd0-ImageDescription = test;keyword
After reload exif-ifd0-ImageDescription = Title (Title, ASCII, 6 components, 6 bytes)
```

It does string load, I'll just add string save. Thanks!

**Comment by jcupitt** [comment]
Enhancement issue for this: https://github.com/jcupitt/libvips/issues/1025

**Comment by jcupitt** [comment]
... I see why this was never implemented, setting string tags with libexif is awful. Sample code:

https://github.com/libexif/libexif/blob/master/contrib/examples/write-exif.c#L228

What a mess!

**Comment by jcupitt** [comment]
Here's a branch with string write implemented: https://github.com/jcupitt/libvips/tree/add-exif-string-write

It seems to work, but probably needs some more testing.

```
john@yingna ~/try $ ./exif2.py ~/pics/exif-test.jpg 
Old exif-ifd0-XPComment = Comment (Comment, Byte, 16 components, 16 bytes)
New exif-ifd0-XPComment = test;keyword
After reload exif-ifd0-XPComment = test;keyword (test;keyword, Byte, 24 components, 24 bytes)
```

**Comment by kleisauke** [comment]
I just tested it and it seems to work on all Microsoft EXIF fields! 👍 

Setting the `exif-ifd0-ImageDescription` tag is broken, though. See:
```
Old exif-ifd0-ImageDescription: Communications (Communications, ASCII, 15 components, 15 bytes)
New exif-ifd0-ImageDescription: test;keyword

(NetVips:2692): VIPS-WARNING **: bad exif meta "exif-ifd3-GPSVersionID"
After reload exif-ifd0-ImageDescription: 20 bytes undefined data (20 bytes undefined data, Undefined, 20 components, 20 bytes)
```
(test image can be found [here](https://github.com/drewnoakes/metadata-extractor-images/blob/master/jpg/FujiFilm%20FinePixS1Pro%20(1).jpg))

@goarsna If you want, you can test this enhancement with:
```csharp
const string tag = "exif-ifd0-XPKeywords";

var image = Image.NewFromFile("test.jpg"); // Load image

Console.WriteLine($"Old {tag}: {image.Get(tag)}");

// libvips images are immutable, so you can't just change exif data,
// you need to make a new image with those values changed.
// To set a new value, use .Copy().
var newImage = image.Copy();
newImage.Set(tag, "test;keyword"); // Change EXIF data

Console.WriteLine($"New {tag}: {newImage.Get(tag)}");

newImage.WriteToFile("result.jpg"); // Write to file

var reloadedImage = Image.NewFromFile("result.jpg"); // Reload file
Console.WriteLine($"After reload {tag}: {reloadedImage.Get(tag)}");
```

The pre-compiled libvips Windows binary with this enhancement can be downloaded from here:
https://t0.nl/libvips/v8.7.0/vips-dev-w64-web-8.7.0-exif-string-write.zip

**Comment by jcupitt** [comment]
OK, I think I've fixed `ImageDescription`. 

libexif has three different types of string write :( yuk!

**Comment by goarsna** [comment]
First of all: Thank you for your quick responses and your work. :)

@kleisauke I downloaded the above provided test image "FujiFilm FinePixS1Pro (1).jpg" and I could reproduce your error. After reloading it complains about x bytes of undefined data in the ImageDescription field.

@jcupitt @kleisauke Though with the new pre-compiled version of libvips I always get the VIPS-WARNING about the GPSVersionID field, even if I just load an image and save that image again.
```
var image = Image.NewFromFile("test.jpg"); // Load image
image.WriteToFile("result.jpg"); // Write to file
```
with any file containing exif data gives the warning
`(NetVips:12388): VIPS-WARNING **: bad exif meta "exif-ifd3-GPSVersionID"`

@jcupitt @kleisauke In all other cases (except from setting the ImageDescription field) your fix provided by the pre-combiled libvips 8.7.0 binary seems to work for setting EXIF string fields.

**Comment by kleisauke** [comment]
Just tested https://github.com/jcupitt/libvips/commit/63bb38056963555627737e84b2328cb9f183dc05 with this test case:
<details>
 <summary>Test case</summary>

```csharp
[SkippableFact]
public void TestExifStringOverride()
{
    Skip.IfNot(Helper.Have("jpegload") && Base.AtLeastLibvips(8, 7), "requires libvips >= 8.7");

    string[] testFields =
    {
        // ASCII with encoding
        "exif-ifd2-UserComment",
        // plain ASCII
        "exif-ifd0-Make",
        "exif-ifd0-Model",
        "exif-ifd0-ImageDescription",
        "exif-ifd0-Artist",
        // UTF16
        "exif-ifd0-XPTitle",
        "exif-ifd0-XPComment",
        "exif-ifd0-XPAuthor",
        "exif-ifd0-XPKeywords",
        "exif-ifd0-XPSubject"
    };

    var x = Image.NewFromFile(Helper.JpegFile);
    var imageFields = x.GetFields();

    foreach (var testField in testFields)
    {
        if (Array.IndexOf(imageFields, testField) >= 0)
        {
            x.Set(testField, "override");
        }
        else
        {
            x.SetType(GValue.GStrType, testField, "override");
        }
    }

    var buf = x.WriteToBuffer(".jpg");
    var y = Image.NewFromBuffer(buf);

    foreach (var testField in testFields)
    {
        Assert.StartsWith("override", (string) y.Get(testField));
    }
}
```
</details>

And it successfully passes! :tada: 

The updated pre-compiled libvips Windows binary can be downloaded from [here](https://t0.nl/libvips/v8.7.0/vips-dev-w64-web-8.7.0-exif-string-write.zip).

@goarsna The `GPSVersionID` warning is indeed annoying. You could try to define the environment variable `VIPS_WARNING` to hide warning messages from libvips.

**Comment by jcupitt** [comment]
That's great!

I'm sure there are more of these string valued tags, those were just the few I spotted immediately.

The date tags also need special handling, I think :( argh. I'll wait for someone to complain about that though.

**Comment by goarsna** [comment]
Once again thanks, guys, for your work!

@kleisauke  The `GPSVersionID` isn't  a problem for me (except that, as you both said' it's annoying). I just mentioned it before, because I only noticed it after libvips update. But thank you for the hint with the environment variable.

@jcupitt Often it's the best option to just wait ;) :D

**Comment by jcupitt** [comment]
OK, merged to master, this improvement will be in libvips 8.7. Thanks for pointing out the dumbness!

**Comment by kleisauke** [comment]
I've just released NetVips 1.0.5 which bundle and depend upon libvips 8.7.
